lib/service-worker.js

```diff
@@ -0,0 +1,103 @@
+'use strict';
+
+/**
+ * ServiceWorker handler and capturer
+ */
+
+const _parseLinkHeader = require('parse-link-header');
+
+const CSS_KEY = 'CSS_KEY';
+const JS_KEY = 'JS_KEY';
+
+module.exports = {
+    handler,
+    captureContent
+};
+
+function captureContent(store, link) {
+    const links = _parseLinkHeader(link);
+    const css = [], js = [];
+    if (links.stylesheet && links.stylesheet.url) {
+        css.push(links.stylesheet.url);
+    }
+    if (links['fragment-script'] && links['fragment-script'].url) {
+        js.push(links['fragment-script'].url);
+    }
+    return saveContent(store, { css, js });
+}
+
+function saveContent(store, data) {
+    const update = {
+        css: actual => store.set(
+            CSS_KEY,
+            [...new Set([...actual, ...data.css])].sort()
+        ),
+        js: actual => store.set(
+            JS_KEY,
+            [...new Set([...actual, ...data.js])].sort()
+        )
+    };
+
+    const createOrReplace = {
+        css: store.get(CSS_KEY)
+            .then(
+                actual => update.css(actual),
+                () => update.css([])
+            ),
+        js: store.get(JS_KEY)
+            .then(
+                actual => update.js(actual),
+                () => update.js([])
+            )
+    };
+
+    return Promise.all([
+        createOrReplace.css, createOrReplace.js
+    ]);
+}
+
+function handler(options, request, response) {
+    if (!isObject(options.serviceWorker)) {
+        return respondNotEnabled(response);
+    }
+
+    response.writeHead(200, {
+        'Content-Type': 'application/javascript'
+    });
+
+    let cssVarName, jsVarName;
+    if (isObject(options.serviceWorker.varNames)) {
+        cssVarName = options.serviceWorker.varNames.css
+            ? options.serviceWorker.varNames.css
+            : 'CSS';
+        jsVarName = options.serviceWorker.varNames.js
+            ? options.serviceWorker.varNames.js
+            : 'JS';
+    } else {
+        [cssVarName, jsVarName] = ['CSS', 'JS'];
+    }
+
+    const {store} = options;
+
+    return Promise.all([
+        store.get(CSS_KEY),
+        store.get(JS_KEY)
+    ]).then(([css, js]) => {
+        response.end(`
+            var ${cssVarName} = [${css.map(c => JSON.stringify(c))}];
+            var ${jsVarName} = [${js.map(c => JSON.stringify(c))}];
+            ${options.serviceWorker.contents}
+        `);
+    });
+}
+
+function isObject(o) {
+    return !Array.isArray(o) && typeof o === 'object' && o !== null;
+}
+
+function respondNotEnabled(response) {
```

Comment:
would move this function definition to before it gets used. relying on hoisting can confuse.